rfcs/0000-type-stability.md

```diff
@@ -0,0 +1,500 @@
+
+[wider-property]: https://www.typescriptlang.org/play/#code/CYUwxgNghgTiAEkoGdnwPIwJYHMsDsoJ4BvAKHnjimAHt8IBPeABxlpZBgBdGAueMm7Z8OANxkAvmTKgkcRNFTwA6llD4QwUhSoga9Jq3ace-QcII54AHwsicAbQC6UmQDMArvjDcs9eAALKHxgCBBkTFwCIgA5WHYAdwAKADciTxABKLxCCABKAXxPAFsAIy4dSjhuTxh8eHSITIA6Ng4uXhbw0W5AiWlZcGgFcO54KGzsXKIJORGEMfgygTUNLQlg0PDI6ZiIeJgk5Kh8zZCwiJz9w+Oys7IgA
+
+[writable-property]: https://www.typescriptlang.org/play/#code/PTAEFpNB3BLATApgO1sg5qARgJ0QQwGsBnCKEAWACg0AXRHAM3wGNFQB1WAG266WSgA3tVBjQ+AFyhitHGnQBuagF9q1Og2ZtQAEQT8Uw0eKky5C0AB9QyAK4BbLA2VU1VaozvIWtWAHtBAAt8ZHhuRGIuXkNkAAoAN3xuO0RpaL4EFABKaXsnBmMqcVA8WjscQSSUxAA6fFqIjFog13dqJBZufDxQCNoYHkyBdKHY1xCwiKixrPi4GLns1w7ELp72ftB4AznpfXhx6knwyIzYuJ3DpZWqEDJwWx6cfzgMCW5if2w8IlJIR6UDw0ZD0JisdgZAByz1eRRKZlk8neNnyzhwbXUILB2nYBxhOBe0HhpjyjnRmOBXh8fkCMHk9GIABV-NDYdBEslUqNeASidkSWJqql6qAALygADkQUQvH8kspq3WvS2C24fNePPV7Nc0AZkRZbMJrziao10GWWM63RViAGV3N+wQ5t1+uZrKG5suzvZlqoQA
+
+Each of these will trigger a compiler error for consumers, surfacing the error. As such, they should be easily detectable by testing infrastructure (see below under [Tooling: Detect breaking changes in types](#detect-breaking-changes-in-types)).
+
+There are several reasons why breaking changes may occur:
+
+-   The author of the addon may choose to change the API for whatever reason. This is no different than the situation today for addons which do not support TypeScript.
+
+-   The author of the addon may need to make changes to adapt to changes in Ember, for example to support Octane idioms. This is likewise identical with the situation for addons which do not support TypeScript.
+
+-   Adopting a new version of TypeScript may change the meaning of existing types. For example, in TypeScript 3.5, generic types without a specified default type changed their default value from `{}` to `unknown`. This improved type safety, but [broke many existing types][3.5-breakage], including the internal types for Glimmer and Ember.
+
+-   Adopting a new version of TypeScript may change the type definitions emitted in `.d.ts` files in backwards-incompatible ways. For example, changing to use the finalized ECMAScript spec for class fields meant that [types emitted by TypeScript 3.7 were incompatible with TypeScript 3.5 and earlier][3.7-emit-change].
+
+[3.5-breakage]: https://github.com/microsoft/TypeScript/issues/33272
+[3.7-emit-change]: https://github.com/microsoft/TypeScript/pull/33470
+
+#### Non-breaking changes to type definitions
+
+The following are *not* breaking changes:
+
+-   a function (including a class method or constructor) *accepts a less specific (“wider”) type*, for example if it previously accepted only a `boolean` but now accepts `boolean | undefined`—since all existing user code will continue working and type-checking ([playground][wider-argument])
```

Comment:
I think the same distinction applies here as above—if the widening accurately describes the runtime behavior, then it's a non-breaking change. If it doesn't, it's a bug and should be fixed in a patch release.

Reply:
Agreed with @dfreeman here as well. If the *library* changes its behavior to now accept `any` or `unknown`, then that widening is *not* a breaking change for any consumer. By contrast, if the types for a function are changed to say that an argument now can be `any`, the error is not that they're wider than they were previously but that they're wider than, and therefore do not correctly represent, the *runtime* behavior.

I will add a note to the beginning of this whole discussion noting that for the purposes of this discussion, we're assuming that the narrowing and widening of types discussed throughout is *in keeping with the runtime API*, because this is a very important point.

Reply:
I've added both a note to the introduction and a longer discussion under the definition of bug fixes to account for this!